Improve documentation in dispatcher; remove unnecessary optional

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32782 Replace catchall kernels with CompoundOp dispatch key.
* #32787 Stop using dispatchTypeId to do checks for tensor list unwrap.
* #32734 Centralize addition of "always on" dispatch keys.
* #32729 Make DispatchKeyGuards accept DispatchKey::Undefined
* #32728 Rename DispatchKey::UndefinedTensorId to Undefined
* #32727 Delete copy/move constructors on these RAII guards.
* #32557 Add missing C10_API to dispatch key TLS setter/getters
* **#32533 Improve documentation in dispatcher; remove unnecessary optional**

Applies renames based on comments in #32439.  I also updated some
other documentation and variable names while I was at it.

Fixes #32435.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19579854](https://our.internmc.facebook.com/intern/diff/D19579854)

ezyang committed:
Improve documentation in dispatcher; remove unnecessary optional

Applies renames based on comments in #32439.  I also updated some
other documentation and variable names while I was at it.

Fixes #32435.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 6e114859:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6e1148598596c7d164e43182490f64b16cc0623e).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4325762) pytorch_linux_xenial_cuda10_1_cudnn7_py3_multigpu_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89950189) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89950189)) :snowflake:
<details>
<summary>
<code>Jan 23 19:29:03 AssertionError: 10 not less than or equal to 1e-05 :</code>
</summary>

```
Jan 23 19:29:03 ---------------------------------------------------------------------- 
Jan 23 19:29:03 Traceback (most recent call last): 
Jan 23 19:29:03   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 174, in wrapper 
Jan 23 19:29:03     self._join_processes(fn) 
Jan 23 19:29:03   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 255, in _join_processes 
Jan 23 19:29:03     self._check_return_codes(elapsed_time) 
Jan 23 19:29:03   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 275, in _check_return_codes 
Jan 23 19:29:03     self.assertEqual(p.exitcode, first_process.exitcode) 
Jan 23 19:29:03   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 892, in assertEqual 
Jan 23 19:29:03     super(TestCase, self).assertLessEqual(abs(x - y), prec, message) 
Jan 23 19:29:03 AssertionError: 10 not less than or equal to 1e-05 :  
Jan 23 19:29:03  
Jan 23 19:29:03 ---------------------------------------------------------------------- 
Jan 23 19:29:03 Ran 117 tests in 207.208s 
Jan 23 19:29:03  
Jan 23 19:29:03 FAILED (failures=1, skipped=4) 
Jan 23 19:29:03  
Jan 23 19:29:03 Generating XML reports... 
Jan 23 19:29:04 Traceback (most recent call last): 
Jan 23 19:29:04   File "test/run_test.py", line 456, in <module> 
Jan 23 19:29:04     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -68,57 +66,73 @@ namespace detail {
-  // docs on nonFallthroughKernels_
-  void setIsOperatorOverridden(DispatchKey k, bool is_overridden);
+  // docs on operatorHasKernelForBackend_
+  void setOperatorHasKernelForBackend(DispatchKey k, bool is_overridden);
```
`bool is_overridden` should probably also be renamed

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -68,57 +66,73 @@ namespace detail {
+    // We must NOT respect the passed in backendsWithoutFallthrough if an operator has
+    // specifically overridden the backend, since that means we've opted to
+    // not fallthrough and instead apply some specific behavior (which we
+    // must dispatch to).  For now, we assume that operators NEVER are the
```
"operators NEVER override the backend with a fallthrough kernel"

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -144,8 +144,10 @@ class CAFFE2_API Dispatcher final {
-  DispatchKeySet backendFallbackNonFallthroughSet_;
+  // Set of backends which have specified they do NOT want fallthrough behavior
+  // (we store the inverse because it avoids a negation when we use this for
+  // maskign)
```
typo

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -208,31 +210,28 @@ template<class Return, class... Args>
 inline Return Dispatcher::callUnboxed(const OperatorHandle& op, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
   const auto& dispatchTable = op.operatorIterator_->op.dispatch_table();
-  c10::optional<DispatchKey> dispatchKey = dispatchTable.dispatchKeyExtractor().getDispatchKeyUnboxed<Args...>(backendFallbackNonFallthroughSet_, args...);
```
We should probably also change the `callUnboxedWithDispatchKey` signature to not take an optional anymore

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -208,31 +210,28 @@ template<class Return, class... Args>
 
 inline const KernelFunction& Dispatcher::dispatch_(const DispatchTable& dispatchTable, c10::optional<DispatchKey> dispatchKey) const {
-  if (C10_LIKELY(dispatchKey.has_value())) {
+  const KernelFunction* backendKernel = dispatchTable.lookup(*dispatchKey);
```
As long as the parameter is an optional, you can't just unconditionally dereference it. But as stated above, we should probably change this signature here too to take a non-optional DispatchKey and then this is fine.

smessmer reviewed (approved):
Thanks for doing these changes. I find it much clearer than before.

ezyang commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -11,34 +11,32 @@ namespace c10 {
+    // for consideration by the included TLS), which is why you have to pass
+    // them in to this function (as opposed to just applying it to the input
+    // 'ts').
+    DispatchKeySet backendsWithoutFallthrough
```
New wording:

```
      // The key mask lets us eliminate (by zero entries) keys which should not
      // be considered for dispatch.  There is one case when we use this:
      // if there is no operator registered for a backend whose fallback behavior
      // is to fallthrough, we eliminate that backend from consideration (since
      // we want to "fallthrough" to the next valid key.)
      //
      // These excluded backends are NOT tracked in the TLS, but must be applied
      // AFTER TLS (since the backend may have been introduced for consideration
      // by the included TLS), which is why you have to pass them in to this
      // function (as opposed to just applying it to the input 'ks').
      DispatchKeySet key_mask
```

ezyang committed:
Update on "Improve documentation in dispatcher; remove unnecessary optional"


Applies renames based on comments in #32439.  I also updated some
other documentation and variable names while I was at it.

Fixes #32435.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19579854](https://our.internmc.facebook.com/intern/diff/D19579854)

[ghstack-poisoned]

ezyang committed:
Update on "Improve documentation in dispatcher; remove unnecessary optional"


Applies renames based on comments in #32439.  I also updated some
other documentation and variable names while I was at it.

Fixes #32435.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19579854](https://our.internmc.facebook.com/intern/diff/D19579854)

[ghstack-poisoned]

ezyang committed:
Update on "Improve documentation in dispatcher; remove unnecessary optional"


Applies renames based on comments in #32439.  I also updated some
other documentation and variable names while I was at it.

Fixes #32435.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19579854](https://our.internmc.facebook.com/intern/diff/D19579854)

[ghstack-poisoned]
